x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CreateAndFollowIndexAction.java

```diff
@@ -0,0 +1,309 @@
+            if (leaderIndexMetaData == null) {
+                listener.onFailure(new IllegalArgumentException("leader index [" + request.getFollowRequest().getLeaderIndex() +
+                    "] does not exist"));
+                return;
+            }
+
+            ActionListener<Boolean> handler = ActionListener.wrap(
+                result -> {
+                    if (result) {
+                        initiateFollowing(request, listener);
+                    } else {
+                        listener.onResponse(new Response(true, false, false));
+                    }
+                },
+                listener::onFailure);
+            clusterService.submitStateUpdateTask("follow_index_action", new AckedClusterStateUpdateTask<Boolean>(request, handler) {
+
+                @Override
+                protected Boolean newResponse(boolean acknowledged) {
+                    return acknowledged;
+                }
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    IndexMetaData currentIndex = currentState.metaData().index(request.getFollowRequest().getFollowIndex());
+                    if (currentIndex != null) {
+                        throw new ResourceAlreadyExistsException(currentIndex.getIndex());
+                    }
+
+                    MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
+                    if (request.getFollowRequest().getFollowIndex() != null) {
+                        IndexMetaData.Builder imdBuilder = IndexMetaData.builder(leaderIndexMetaData);
+
+                        Settings.Builder settingsBuilder = Settings.builder();
+                        settingsBuilder.put(leaderIndexMetaData.getSettings());
+                        // TODO: do we want leader and follow index to have the same UUID?
+                        // (Overwriting UUID here, because otherwise we can't follow indices in the same cluster)
+                        settingsBuilder.put(IndexMetaData.SETTING_INDEX_UUID, UUIDs.randomBase64UUID());
+                        settingsBuilder.put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowIndex());
+                        settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+
+                        imdBuilder.settings(settingsBuilder);
+                        imdBuilder.index(request.getFollowRequest().getFollowIndex());
+                        for (int shardId = 0 ; shardId < leaderIndexMetaData.getNumberOfShards(); shardId++) {
+                            imdBuilder.putInSyncAllocationIds(shardId, new HashSet<>());
+                        }
+                        mdBuilder.put(imdBuilder.build(), false);
+                    } else {
+                        mdBuilder.put(leaderIndexMetaData, false);
```

Comment:
I don't think this case is possible; what am I missing?

Reply:
True, this is something that I should have removed.